audio/pwaudio.c

```diff
@@ -0,0 +1,776 @@
+		    SPA_AUDIO_FORMAT_S32_LE;
+		break;
+	case AUDIO_FORMAT_U32:
+		format =
+		    endianness ? SPA_AUDIO_FORMAT_U32_BE :
+		    SPA_AUDIO_FORMAT_U32_LE;
+		break;
+	case AUDIO_FORMAT_F32:
+		format =
+		    endianness ? SPA_AUDIO_FORMAT_F32_BE :
+		    SPA_AUDIO_FORMAT_F32_LE;
+		break;
+	default:
+		dolog("Internal logic error: Bad audio format %d\n", fmt);
+		format = SPA_AUDIO_FORMAT_U8;
+		break;
+	}
+	return format;
+}
+
+static AudioFormat pw_to_audfmt(enum spa_audio_format fmt, int *endianness,
+				uint32_t * frame_size)
+{
+	switch (fmt) {
+	case SPA_AUDIO_FORMAT_S8:
+		*frame_size = 1;
+		return AUDIO_FORMAT_S8;
+	case SPA_AUDIO_FORMAT_U8:
+		*frame_size = 1;
+		return AUDIO_FORMAT_U8;
+	case SPA_AUDIO_FORMAT_S16_BE:
+		*frame_size = 2;
+		*endianness = 1;
+		return AUDIO_FORMAT_S16;
+	case SPA_AUDIO_FORMAT_S16_LE:
+		*frame_size = 2;
+		*endianness = 0;
+		return AUDIO_FORMAT_S16;
+	case SPA_AUDIO_FORMAT_U16_BE:
+		*frame_size = 2;
+		*endianness = 1;
+		return AUDIO_FORMAT_U16;
+	case SPA_AUDIO_FORMAT_U16_LE:
+		*frame_size = 2;
+		*endianness = 0;
+		return AUDIO_FORMAT_U16;
+	case SPA_AUDIO_FORMAT_S32_BE:
+		*frame_size = 4;
+		*endianness = 1;
+		return AUDIO_FORMAT_S32;
+	case SPA_AUDIO_FORMAT_S32_LE:
+		*frame_size = 4;
+		*endianness = 0;
+		return AUDIO_FORMAT_S32;
+	case SPA_AUDIO_FORMAT_U32_BE:
+		*frame_size = 4;
+		*endianness = 1;
+		return AUDIO_FORMAT_U32;
+	case SPA_AUDIO_FORMAT_U32_LE:
+		*frame_size = 4;
+		*endianness = 0;
+		return AUDIO_FORMAT_U32;
+	case SPA_AUDIO_FORMAT_F32_BE:
+		*frame_size = 4;
+		*endianness = 1;
+		return AUDIO_FORMAT_F32;
+	case SPA_AUDIO_FORMAT_F32_LE:
+		*frame_size = 4;
+		*endianness = 0;
+		return AUDIO_FORMAT_F32;
+	default:
+		*frame_size = 1;
+		dolog("Internal logic error: Bad spa_audio_format %d\n", fmt);
+		return AUDIO_FORMAT_U8;
+	}
+}
+
+static int create_stream(pwaudio * c, PWVoice * v, const char *name)
+{
+	int res;
+	uint32_t n_params;
+	const struct spa_pod *params[2];
+	uint8_t buffer[1024];
+	struct spa_pod_builder b;
+
+	v->stream = pw_stream_new(c->core, name, NULL);
+
+	if (v->stream == NULL) {
+		res = -errno;
+		goto error_free;
+	}
+
+	if (v->mode == MODE_SOURCE) {
+		pw_stream_add_listener(v->stream,
+				       &v->stream_listener,
+				       &capture_stream_events, v);
+	} else {
+		pw_stream_add_listener(v->stream,
+				       &v->stream_listener,
+				       &playback_stream_events, v);
+	}
+
+	n_params = 0;
+	spa_pod_builder_init(&b, buffer, sizeof(buffer));
+	params[n_params++] = spa_format_audio_raw_build(&b,
+							SPA_PARAM_EnumFormat,
+							&v->info);
+
+	//connect the stream to a sink and source
+	if ((res = pw_stream_connect(v->stream,
+				     v->mode ==
+				     MODE_SOURCE ? PW_DIRECTION_INPUT :
+				     PW_DIRECTION_OUTPUT, PW_ID_ANY,
+				     PW_STREAM_FLAG_AUTOCONNECT |
+				     PW_STREAM_FLAG_MAP_BUFFERS |
+				     PW_STREAM_FLAG_RT_PROCESS, params,
+				     n_params)) < 0)
+		return res;
```

Comment:
Omitting the curly brackets is generally more error-prone. 

Also, as you have defined the label, maybe `goto error_free;` instead?